lib/tlscontext.c

```diff
@@ -1092,10 +1140,23 @@ tls_context_new(TLSMode mode, const gchar *location)
 }
 
 static void
-_tls_context_free(TLSContext *self)
+_tls_ssl_context_free(TLSContext *self)
 {
-  g_free(self->location);
   SSL_CTX_free(self->ssl_ctx);
+  self->ssl_ctx = NULL;
+#if SYSLOG_NG_HAVE_DECL_SSL_CONF_CMD
+  SSL_CONF_CTX_free(self->ssl_conf_ctx);
+  self->ssl_conf_ctx = NULL;
+  g_list_foreach(self->conf_cmds_list, (GFunc)g_free, NULL);
```

Comment:
the cmd list should probably be freed in tls_context_free() and not here. we only want to free it once (a single TLSContext instance can create a new SSL instance multiple times.

Reply:
This call is part of _tls_ssl_context_free() that
- was born to have a common internal cleanup function (instead of the earlier code repetition) 
- is called basically 
  - from _tls_context_free() as you wrote, and
  - when we do initialization error cleanup (but those are exceptional cases) 

Also, the ssl_conf_ctx is bound to the ssl_conf_ctx (via SSL_CONF_CTX_set_ssl_ctx()) so I think they should be treated as one and handled/freed together.

Reply:
I think the command list itself is really a property of TLSContext. There's a method tls_context_set_conf_cmds() that sets them and I think they should remain available until TLSContext itself is freed.

It is TLSContext's responsibility  to set up an SSL_CTX based on settings stored in its members, by invoking tls_context_setup_context()

Although tls_context_setup_context() implementation seems to assume that it can only be invoked once, I still think 
that anything that is needed to setup SSL_CTX should remain to be available. 

We are not freeing cert_file or ca_file either, do we? 

I think we should review the layering here (I also have a parallel PR that touches on this subject, see #4221):
  * TLSContext is an object that stores all tls() related settings,
  * based on those settings it is able to initialize an SSL_CTX
  * based on SSL_CTX it can produce SSL instances

The initialization of SSL_CTX is decoupled from the constructor simply it is not the right time to report errors in most cases.

With that in mind, all settings (parsed from the config) should be stored in TLSContext until the life-time of that object and not freed, especially not partially.

I view tls_context_setup_context() similar to what init() does in the rest of the codebase (we have the init-style conventions in a lot of places in the codebase), init() should not free stuff that could be needed for a repeated init() call.

With all that said, the question is what should be the lifecycle for the SSL_CTX and SSL_CONF_CTX objects. SSL_CTX is currently initialized in the constructor, this would probably be better situated in the setup_context() call. 

With SSL_CONF_CTX, I view that as a temporary, which is _ONLY_ needed for initialization and not any later than that. 
  * so my conclusion would be to make SSL_CONF_CTX just a local variable in tls_context_setup_cmd_context()
  * and maybe move the creation of ssl_ctx to the setup_context() call (and free it and only it if there's an error)

This means that TLSContext->ssl_ctx being NULL clearly indicates an error: either that the setup was unsuccessful or that TLSContext wasn't even initialized.

Does this make sense?

Reply:
I see your point, but I guess such a refactor should go to a separate run, I guess the ongoing #4221 is a good candidate, can we move it over there?

Reply:
It must be my fault, sorry, but after reading over it once again, I can see now, we are not on the same page either ;)

What I see now (and it is the reason I implemented this way)

1. As you mentioned it is not clear how the ssl_ctx bound ssl_conf_ctx should be handled. In my interpretation, once it is bound to the ssl_ctx it is part of it and must be handled and freed together with it.
2. This is the main reason.
There's nothing in the code that prevents a given syslog-ng.conf TLS section item (same is true for other sections and items as well) to appear repeatedly.
This can be treated as a feature or a bug, I think it is a useful feature.
According to this, we have to be prepared for this case, this is why I've added 
- missing resource freeing code (more might exist, needs further investigation) in some functions that are called at the end of parsing of a given section item (e.g.  tls_session_set_trusted_dn)
- prepared setup of ssl_conf_ctx to be able to handle repeated setup calls
3. 
 ```
/* freeing of ssl_ctx and setting it to NULL should be done at the error label in tls_context_setup_context(), this should be just one location and not more. */
  SSL_CTX_free(self->ssl_ctx);
  self->ssl_ctx = NULL;
```
Exactly. It should have been, but it has not been.
In the original version, you can see there were 2 different kinds of error handling, one that cleared up the ssl_ctx without calling `_print_and_clear_tls_session_error(self);` and one with it (via goto error;)
That is why _tls_ssl_context_free() was born 
- to keep the original logic (not to call `_print_and_clear_tls_session_error(self);` when it is not needed), and
- to replace the otherwise repeated code
              SSL_CTX_free(self->ssl_ctx);
              self->ssl_ctx = NULL;
- to create a function that could handle the management of the ssl_ctx bound ssl_conf_ctx as well

Hope this makes sense.

Reply:
My assumption was that the SSL_CONF_ctx is not needed once the commands have been executed, so you can free that right after that. It is only bound to a specific ssl_ctx as it is making changes to it due to the commands.

Reply:
the actual implementation handles nicely the possibly repeated config entries, I think we should keep this behavior

Reply:
@HofiOne 

> In my interpretation, once it is bound to the ssl_ctx it is part of it and must be handled and freed together with it.

My understanding is that this statement is not true.
`SSL_CONF_CTX` can be considered as a configurator object that calls SSL and SSL_CTX API calls based on strings you provide using `SSL_CONF_cmd()`.
After you send in the last config command, the `SSL_CONF_CTX` can be freed.

This is documented behavior, even if somewhat implicitly stated.

An example of this "local" usage:
https://github.com/nginx/nginx/blob/765ef1098d169622ada67eb4eccad1d45d47a61c/src/event/ngx_event_openssl.c#L1575-L1620

Reply:
I Hope fixed all the discussed topics now.